pkg/sql/catalog/randgen/templates.go

```diff
@@ -0,0 +1,220 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package randgen
+
+import (
+	"context"
+	"sort"
+
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/catenumpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/parser"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/types"
+	"github.com/cockroachdb/errors"
+)
+
+type tbTemplate struct {
+	namePat         string
+	desc            descpb.TableDescriptor
+	baseColumnNames []string
+}
+
+func (g *testSchemaGenerator) loadTemplates(ctx context.Context) {
+	if !g.createTables {
+		// No template needed.
+		return
+	}
+	if len(g.cfg.TableTemplates) == 0 {
+		// No template specified: we use a simple predefined table template.
+		g.templates = append(g.templates, defaultTemplate())
+		return
+	}
+
+	// The user has specified some table patterns. Look them up.
+	objIDs := make(map[descpb.ID]struct{})
+outer:
+	for _, pat := range g.cfg.TableTemplates {
+		// The user may specify a template either as an optionally
+		// qualified table name, or a "table name pattern": a name whose
+		// last component is a '*'.
+		tbPat, err := parser.ParseTablePattern(pat)
+		if err != nil {
+			panic(genError{errors.Wrapf(err, "parsing template name %q", pat)})
+		}
+
+		// The pattern -> table IDs expansion is provided by the caller.
+		// We share this code with the GRANT statement.
+		_, ids, err := g.cat.ExpandTableGlob(ctx, tbPat)
+		if err != nil {
+			panic(genError{errors.Wrapf(err, "expanding template name %q", pat)})
+		}
+
+		// De-dup the objects: if a user specifies e.g.
+		// ["foo.*","foo.bar"] we should use foo.bar only once.
+		for _, id := range ids {
+			const maxTemplates = 100
+			if len(objIDs) > maxTemplates {
+				// Let's not let the template list get out of hand.
+				break outer
+			}
+			objIDs[id] = struct{}{}
+		}
+	}
+
+	// We want name generation and template reuse to be deterministic as
+	// a function of the random seed, so we must remove non-determinism
+	// coming from the go map order.
+	sobjIDs := make([]descpb.ID, 0, len(objIDs))
+	for id := range objIDs {
+		sobjIDs = append(sobjIDs, id)
+	}
+	sort.Slice(sobjIDs, func(i, j int) bool { return sobjIDs[i] < sobjIDs[j] })
+
+	// Look up the descriptors from the IDs.
+	descs, err := g.coll.GetImmutableDescriptorsByID(ctx, g.txn,
+		tree.CommonLookupFlags{Required: true},
```

Comment:
nit: `Required: true` is always implied by `GetImmutableDescriptorsByID`. One of the many sad little warts of the collection API which I hope to get rid of soon, see https://github.com/cockroachdb/cockroach/pull/93813 if you're interested in this.